api-server/llama-core/src/search.rs

```diff
@@ -0,0 +1,389 @@
+    pub content_type: ContentType,
+    /// The (expected) content type of the output.
+    pub output_content_type: ContentType,
+    /// Method expected by the api endpoint.
+    pub method: String,
+    /// Additional headers for any other purpose.
+    pub additional_headers: Option<std::collections::HashMap<String, String>>,
+    /// Callback function to parse the output of the api-service. Implementation left to the user.
+    pub parser: fn(&serde_json::Value) -> Result<SearchOutput, Box<dyn std::error::Error>>,
+    /// Prompts for use with summarization functionality. If set to `None`, use hard-coded prompts.
+    pub summarization_prompts: Option<(String, String)>,
+    /// Context size for summary generation. If `None`, will use the 4 char ~ 1 token metric to generate summary.
+    pub summarize_ctx_size: Option<usize>,
+}
+
+// output format for individual results in the final output.
+#[derive(Serialize, Deserialize)]
+pub struct SearchResult {
+    pub url: String,
+    pub site_name: String,
+    pub text_content: String,
+}
+
+// Final output format for consumption by the LLM.
+#[derive(Serialize, Deserialize)]
+pub struct SearchOutput {
+    pub results: Vec<SearchResult>,
+}
+
+impl SearchConfig {
+    /// Wrapper for the parser() function.
+    pub fn parse_into_results(
+        &self,
+        raw_results: &serde_json::Value,
+    ) -> Result<SearchOutput, Box<dyn std::error::Error>> {
+        (self.parser)(raw_results)
+    }
+    pub fn new(
+        search_engine: String,
+        max_search_results: u8,
+        size_limit_per_result: u16,
+        endpoint: String,
+        content_type: ContentType,
+        output_content_type: ContentType,
+        method: String,
+        additional_headers: Option<std::collections::HashMap<String, String>>,
+        parser: fn(&serde_json::Value) -> Result<SearchOutput, Box<dyn std::error::Error>>,
+        summarization_prompts: Option<(String, String)>,
+        summarize_ctx_size: Option<usize>,
+    ) -> SearchConfig {
+        SearchConfig {
+            search_engine,
+            max_search_results,
+            size_limit_per_result,
+            endpoint,
+            content_type,
+            output_content_type,
+            method,
+            additional_headers,
+            parser,
+            summarization_prompts,
+            summarize_ctx_size,
+        }
+    }
+    /// Perform a web search with a `Serialize`-able input. The `search_input` is used as is to query the search endpoint.
+    pub async fn perform_search<T: Serialize>(
+        &self,
+        search_input: &T,
+    ) -> Result<SearchOutput, LlamaCoreError> {
+        let client = Client::new();
+        let url = match Url::parse(&self.endpoint) {
+            Ok(url) => url,
+            Err(_) => {
+                let msg = "Malformed endpoint url";
+                #[cfg(feature = "logging")]
+                error!(target: "search", "perform_search: {}", msg);
```

Comment:
Please update the target of all `error!` (and `info!` if present) to `"stdout"` as the entire project has already migrated to `WasmEdge v0.14.x`, which has different `wasi-logging` settings. Thanks!